src/modules/recorder.module.ts

```diff
@@ -0,0 +1,373 @@
+        id += CLIP_ID_CHARS[Math.floor(Math.random() * CLIP_ID_CHARS.length)]
+    }
+    return id
+}
+
+async function saveClip(channel: VoiceChannel, clipName: string, textChannel: TextChannel) {
+    const voiceChannelData = voiceChannels.get(channel.id)
+    if (!voiceChannelData) {
+        return
+    }
+
+    const now = new Date()
+    if (voiceChannelData.lastClipTime.getTime() + CLIP_COOLDOWN_SEC * MILLIS > now.getTime()) {
+        return await textChannel.send("Please wait a few seconds before clipping again.")
+    }
+
+    // Prepare buffer for upload
+    const buffer: Buffer = prepareBuffer(voiceChannelData.buffer, BUFFER_WRITE_SEC)
+    const clipId: string = randomClipId()
+    const name = clipName ? clipName : clipId
+    const params = {
+        Bucket: BUCKET,
+        Key: `${clipId}.${FILE_EXT}`,
+        Body: buffer
+    }
+
+    // Upload file to S3
+    s3.upload(params, (err: Error, data: Record<string, any>) => {
+        if (err) {
+            console.log(err)
+            return
+        }
+
+        // Write entry to database
+        const clip = {
+            id: clipId,
+            time: new Date(),
+            name: name,
+            url: data.Location,
+            clipstart: 0,
+            clipend: buffer.length,
+            participants: channel.members.map(member => member.id)
+        }
+
+        createClip(clip).then(async res => {
+            voiceChannelData.lastClipTime = now
+            await textChannel.send(`Saved clip with name ${name}`)
+        })
+        .catch(err => {
+            // try inserting with name-timestamp
+            clip.name = clip.name + "-" + Math.floor(now.getTime() / MILLIS)
+            createClip(clip).then(async res => {
+                voiceChannelData.lastClipTime = now
+                await textChannel.send(`Saved clip with name ${clip.name}`)
+            }).catch(err => {
+                console.log(err)
+            })
+        })
+    })
+}
+
+async function getClipFromS3(clip: Clip): Promise<Buffer> {
+    const params = {
+        Bucket: BUCKET,
+        Key: `${clip.id}.pcm`,
+        Range: `bytes=${clip.clipstart}-${clip.clipend}`
+    }
+
+    return new Promise<Buffer>((resolve, reject) => {
+        s3.getObject(params, (err, data) => {
+            if (err) {
+                reject(err)
+            } else if (!(data.Body instanceof Buffer)) {
+                reject(data.Body)
+            } else {
+                resolve(data.Body)
+            }
+        })
+    })
+    
+}
+
+async function playClip(channel: VoiceChannel, clipName: string, textChannel: TextChannel) {
+    const voiceChannelData = voiceChannels.get(channel.id)
+    if (!voiceChannelData) {
+        return
+    }
+
+    let clip
+    try {
+        clip = await getClipByName(clipName)
+    } catch (err) {
+        return await textChannel.send("Could not find a clip with that name.")
+    }
+
+    getClipFromS3(clip).then(buffer => {

```

Comment:
I'd use `await` and `try/catch` here, especially as the line above has `async/await`